core/compiler.common.jvm/src/org/jetbrains/kotlin/load/java/AbstractAnnotationTypeQualifierResolver.kt

```diff
@@ -0,0 +1,205 @@
+import org.jetbrains.kotlin.load.java.typeEnhancement.NullabilityQualifier
+import org.jetbrains.kotlin.load.java.typeEnhancement.NullabilityQualifierWithMigrationStatus
+import org.jetbrains.kotlin.name.FqName
+import java.util.concurrent.ConcurrentHashMap
+
+typealias TypeQualifierWithApplicability<Annotation> = Pair<Annotation, Set<AnnotationQualifierApplicabilityType>>
+
+abstract class AbstractAnnotationTypeQualifierResolver<Annotation : Any>(
+    private val javaTypeEnhancementState: JavaTypeEnhancementState
+) {
+    protected abstract val Annotation.annotations: Iterable<Annotation>
+    protected abstract val Annotation.key: Any
+    protected abstract val Annotation.fqName: FqName?
+    protected abstract fun Annotation.enumArguments(onlyValue: Boolean): Iterable<String>
+
+    private fun Annotation.findAnnotation(fqName: FqName): Annotation? =
+        annotations.find { it.fqName == fqName }
+
+    private fun Annotation.hasAnnotation(fqName: FqName): Boolean =
+        annotations.any { it.fqName == fqName }
+
+    private val resolvedNicknames = ConcurrentHashMap<Any, Annotation>()
+
+    fun resolveTypeQualifierAnnotation(annotation: Annotation): Annotation? {
+        if (javaTypeEnhancementState.jsr305.isDisabled) return null
+        if (annotation.fqName in BUILT_IN_TYPE_QUALIFIER_FQ_NAMES || annotation.hasAnnotation(TYPE_QUALIFIER_FQNAME))
+            return annotation
+        if (!annotation.hasAnnotation(TYPE_QUALIFIER_NICKNAME_FQNAME))
+            return null
+        return resolvedNicknames.getOrPut(annotation.key) {
+            // This won't store nulls (ConcurrentHashMap does not permit that), but presumably unless the code
+            // is broken a nickname should be resolvable.
+            annotation.annotations.firstNotNullOfOrNull(::resolveTypeQualifierAnnotation) ?: return null
+        }
+    }
+
+    private fun resolveQualifierBuiltInDefaultAnnotation(annotation: Annotation): JavaDefaultQualifiers? {
+        if (javaTypeEnhancementState.disabledDefaultAnnotations) {
+            return null
+        }
+
+        return BUILT_IN_TYPE_QUALIFIER_DEFAULT_ANNOTATIONS[annotation.fqName]?.let { qualifierForDefaultingAnnotation ->
+            val state = resolveDefaultAnnotationState(annotation).takeIf { it != ReportLevel.IGNORE } ?: return null
+            qualifierForDefaultingAnnotation.copy(
+                nullabilityQualifier = qualifierForDefaultingAnnotation.nullabilityQualifier.copy(isForWarningOnly = state.isWarning)
+            )
+        }
+    }
+
+    private fun resolveDefaultAnnotationState(annotation: Annotation): ReportLevel {
+        val annotationFqname = annotation.fqName
+        if (annotationFqname != null && annotationFqname in JSPECIFY_DEFAULT_ANNOTATIONS) {
+            return javaTypeEnhancementState.getReportLevelForAnnotation(annotationFqname)
+        }
+        return resolveJsr305AnnotationState(annotation)
```

Comment:
BTW is it difficult to resolve jsr-305 annotations through javaTypeEnhancementState.getReportLevelForAnnotation as well?

Reply:
Quite frankly I have no clue; I've not looked much into the parts that don't use FE1.0/FIR-specific classes.

Reply:
Theoretically we can add logic for extracting report level for jsr305 annotation [here](https://github.com/JetBrains/kotlin/blob/4e8654fbb9e43ad5b986815aac869af2f0281fb8/compiler/cli/cli-common/src/org/jetbrains/kotlin/cli/common/arguments/JavaTypeEnhancementStateParser.kt#L60). Then we would get rid of jsr305-specific logic here (and in `JavaTypeEnhancementState`) and use for this `javaTypeEnhancementState.getReportLevelForAnnotation`.

These are just passing thoughts for future improvements.